src/devtools/client/webconsole/utils/autocomplete.ts

```diff
@@ -0,0 +1,245 @@
+import { parseExpression } from "@babel/parser";
+import {
+  isIdentifier as _isIdentifier,
+  isMemberExpression as _isMemberExpression,
+} from "@babel/types";
+const { filter } = require("fuzzaldrin-plus");
+
+type ObjectPreviewProperty = {
+  value: ValueFront;
+  writable: boolean;
+  configurable: boolean;
+  enumerable: boolean;
+  name: string;
+  get?: any;
+  set?: any;
+};
+
+type ObjectPreview = {
+  containerEntries: any[];
+  getterValues: any[];
+  properties: ObjectPreviewProperty[];
+  prototypeId: any;
+  prototypeValue: ValueFront;
+};
+
+type ObjectFront = {
+  className: string;
+  objectId: string;
+  preview?: ObjectPreview;
+};
+
+type ValueFront = {
+  _pause: any;
+  _hasPrimitive: false;
+  _primitive?: boolean;
+  _isBigInt: boolean;
+  _isSymbol: boolean;
+  _object: ObjectFront;
+  _uninitialized: boolean;
+  _unavailable: boolean;
+  _elements: null;
+  _isMapEntry: null;
+};
+
+type Binding = {
+  name: string;
+  value: ValueFront;
+};
+
+type Scope = {
+  bindings: Binding[];
+  actor: string;
+  parent: Scope;
+  functionName?: string;
+  object?: ValueFront;
+  scopeKind: string;
+  type: "block" | string;
+};
+
+// This is used as a property name placeholder for what they user could
+//  be trying to type or find the autocomplete match to.
+const PROPERTY_PLACEHOLDER = "fakeProperty";
+
+// Trims quotation marks from the string. This allows us to match
+// bracket notation properties regardless of whether the user has quotation
+// marks or not. i.e., `foo["ba` vs `foo[ba` will both show "bar" as
+// an autocomplete match.
+function normalizeString(str: string) {
+  return str.toLowerCase().replace(/['"`]+/g, "");
+}
+function fuzzyFilter(candidates: string[], query: string): string[] {
+  if (normalizeString(query) === "") {
+    return candidates;
+  }
+  return filter(candidates, query);
+}
+
+function isIdentifier(expression: string) {
+  try {
+    return _isIdentifier(parseExpression(expression));
+  } catch (e) {
+    return false;
+  }
+}
+function isMemberExpression(expression: string) {
```

Comment:
Actually, the more I stare at this, the more I realize we need to ping @loganfsmyth , because there's no way we're going to do a better job than babel here.

For instance,

`foo["some.long[property`

is a valid thing to try and autocomplete that I think our logic totally chokes on. I'm guessing that using something like `parseExpression` (and maybe chopping off characters from the end until it stops raising for invalid syntax) will tell you: the last valid character (which indicates the kind of access we're trying to do), and also the type of the left-term, which tells us whether we should be trying to autocomplete right now.